src/services/rocksdb/backend.rs

```diff
@@ -0,0 +1,253 @@
+use pin_project::pin_project;
+use rocksdb::TransactionDB;
+
+use crate::adapters::kv;
+use crate::error::new_other_backend_error;
+use crate::Accessor;
+use crate::AccessorCapability;
+use crate::Scheme;
+
+const SCAN_LIMIT: usize = 100;
+
+/// Rocksdb backend builder
+#[derive(Clone, Default, Debug)]
+pub struct Builder {
+    /// The path to the rocksdb database
+    path: Option<PathBuf>,
+    /// the working directory of the Redis service. Can be "/path/to/dir"
+    ///
+    /// default is "/"
+    root: Option<String>,
+}
+
+impl Builder {
+    pub(crate) fn from_iter(it: impl Iterator<Item = (String, String)>) -> Self {
+        let mut builder = Builder::default();
+        for (k, v) in it {
+            let v = v.as_str();
+            match k.as_ref() {
+                "path" => builder.path(v),
+                _ => continue,
+            };
+        }
+        builder
+    }
+
+    /// Set the path to the rocksdb database
+    pub fn path<P: AsRef<Path>>(&mut self, path: P) -> &mut Self {
+        self.path = Some(path.as_ref().into());
+        self
+    }
+
+    /// set the working directory, all operations will be performed under it.
+    ///
+    /// default: "/"
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        if !root.is_empty() {
+            self.root = Some(root.to_owned());
+        }
+        self
+    }
+
+    /// Consumes the builder and returns a `Rocksdb` instance.
+    pub fn build(&mut self) -> Result<impl Accessor> {
+        let path = self.path.take().ok_or_else(|| {
+            new_other_backend_error(
+                HashMap::from([("path".into(), "".into())]),
+                anyhow!("path is required but not set"),
+            )
+        })?;
+        let db = TransactionDB::open_default(&path).map_err(|e| {
+            new_other_backend_error(
+                HashMap::from([("path".into(), path.to_string_lossy().into_owned())]),
+                anyhow!("failed to open the database: {:?}", e),
+            )
+        })?;
+
+        Ok(Backend::new(Adapter { db: Arc::new(db) }))
+    }
+}
+
+/// Backend for rocksdb services.
+pub type Backend = kv::Backend<Adapter>;
+
+#[derive(Clone)]
+pub struct Adapter {
+    db: Arc<TransactionDB>,
+}
+
+impl Debug for Adapter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Adapter");
+        ds.field("path", &self.db.path());
+        ds.finish()
+    }
+}
+
+#[async_trait]
+impl kv::Adapter for Adapter {
+    fn metadata(&self) -> kv::Metadata {
+        kv::Metadata::new(
+            Scheme::Rocksdb,
+            &self.db.path().to_string_lossy(),
+            AccessorCapability::Read | AccessorCapability::Write | AccessorCapability::List,
+        )
+    }
+
+    async fn next_id(&self) -> Result<u64> {
+        let txn = self.db.transaction();
+        match txn.get("next_id").map_err(new_rocksdb_error)? {
+            Some(v) => {
+                let id = String::from_utf8_lossy(&v)
```

Comment:
Maybe we can use [u64::from_be_bytes](https://doc.rust-lang.org/std/primitive.u64.html#method.from_be_bytes) and [u64::to_be_bytes](https://doc.rust-lang.org/std/primitive.u64.html#method.to_be_bytes) to avoid parse into utf-8 unicode.

Reply:
I thought about it. But I'm not sure if it's worth considering to keep the id value human readable 🤔 .

Reply:
Maybe this is only helpful for debugging I think. Will take your suggestion.